src/utransport.rs

```diff
@@ -102,65 +100,71 @@ impl UPClientZenoh {
                     let Some(attachment) = sample.attachment() else {
                         let msg = "Unable to get the attachment".to_string();
                         log::error!("{msg}");
-                        resp_callback(Err(UStatus::fail_with_code(UCode::INTERNAL, msg)));
+                        block_on(
+                            resp_callback.on_error(UStatus::fail_with_code(UCode::INTERNAL, msg)),
+                        );
                         return;
                     };
                     let u_attribute = match UPClientZenoh::attachment_to_uattributes(attachment) {
                         Ok(uattr) => uattr,
                         Err(e) => {
-                            let msg =
-                                format!("Unable to transform attachment to UAttributes: {e:?}");
+                            let msg = format!("Transform attachment to UAttributes failed: {e:?}");
                             log::error!("{msg}");
-                            resp_callback(Err(UStatus::fail_with_code(UCode::INTERNAL, msg)));
+                            block_on(
+                                resp_callback
+                                    .on_error(UStatus::fail_with_code(UCode::INTERNAL, msg)),
+                            );
                             return;
                         }
                     };
                     // Create UMessage
-                    Ok(UMessage {
-                        attributes: Some(u_attribute).into(),
-                        payload: Some(UPayload {
-                            length: Some(0),
-                            format: encoding.into(),
-                            data: Some(Data::Value(sample.payload.contiguous().to_vec())),
+                    block_on(
+                        resp_callback.on_receive(UMessage {
+                            attributes: Some(u_attribute).into(),
+                            payload: Some(UPayload {
+                                length: Some(0),
+                                format: encoding.into(),
+                                data: Some(Data::Value(sample.payload.contiguous().to_vec())),
+                                ..Default::default()
+                            })
+                            .into(),
                             ..Default::default()
-                        })
-                        .into(),
-                        ..Default::default()
-                    })
+                        }),
+                    );
                 }
                 Err(e) => {
                     let msg = format!("Error while parsing Zenoh reply: {e:?}");
                     log::error!("{msg}");
-                    Err(UStatus::fail_with_code(UCode::INTERNAL, msg))
+                    block_on(resp_callback.on_error(UStatus::fail_with_code(UCode::INTERNAL, msg)));
                 }
-            };
-            resp_callback(msg);
+            }
         };
 
         // Send query
```

Comment:
I'm not great at Zenoh :)

I'm trying to understand if there's any issue with us using the `zenoh_key` here alone. Is it possible that if we fired off two requests, one after the other that we'd end up getting them back in the wrong order? Should we make use of the `req_id` at all?

Sorry for newb questions :upside_down_face:

Reply:
Zenoh will take care of the order of queries and not mix them. That's why we don't need to use `req_id` while sending the query.

Reply:
Could you explain a little detail behind how the queries are ordered correctly?

If we sent out two Request messages one immediately after the other, call them A and B. And a Response messages was sent back for B first, how does the ordering get enforced?

If there's some bit of documentation or better yet code / tests that show this I think that'd be great.

Reply:
Each time we create a Zenoh query to send out, there is also a `qid` inside Zenoh, which is used to differentiate different messages. While receiving the response message, Zenoh will use the `qid` to find the corresponding buffer to put the message.

In your case, we called `invoke_method(A)` and `invoke_method(B)`. `invoke_method(A)` should only get the response message from its buffer with the correct `qid`. Response message to B will never go inside the buffer of A.

Since this is the mechanism inside Zenoh, I prefer to hide it instead of exposing to users.

Reply:
Ahhh okay. So even though the `zenoh_key` may be the same, there's the `qid` in the implementation of Zenoh that will correctly tie together each request with the corresponding response. Makes sense.

---------

Maybe not on this PR, but I think it'd be nice to show executing a couple of `invoke_method()` one after the other and showing they come back in the correct order. Ensuring in the test that we `invoke_method()` on another task so we can immediately call it again and show that even though we fired request_A, _request_B and were returned response_B, response_A they were correlated correctly.

Reply:
It's a good suggestion, but I'm not sure how to make the response always in reverse order in the test.
Do you have any thoughts?

Reply:
Hmmm. What about if the Request message contained some identifying information and the UListener impl was set up so that it could use that identifying information to sleep request_A, while not doing so for request_B?

Reply:
Then we should make sure the callback won't block while accepting the first request.
Maybe related to [the issue](https://github.com/eclipse-uprotocol/up-spec/issues/104)

Reply:
@evshary there will be an issue if you do not pass reqid in the zenoh message we will have issues correlating a message across transports. Is the reqid passed in the UMessage returned from the transport??

Reply:
reqid is still passed in UAttributes (in UMessage). It's just not used in Zenoh mechansim.

Reply:
@PLeVasseur Good diagram! I think this is correct.